pageserver/src/tenant/timeline/delete.rs

```diff
@@ -179,6 +176,32 @@ async fn remove_maybe_offloaded_timeline_from_tenant(
     Ok(())
 }
 
+/// It is important that this gets called when DeletionGuard is being held.
+/// For more context see comments in [`DeleteTimelineFlow::prepare`]
+async fn upload_new_tenant_manifest(
+    tenant: &Tenant,
+    _: &DeletionGuard, // using it as a witness
+) -> anyhow::Result<()> {
+    // This is susceptible to race conditions, i.e. we won't continue deletions if there is a crash
+    // between the deletion of the index-part.json and reaching of this code.
+    // So indeed, the tenant manifest might refer to an offloaded timeline which has already been deleted.
+    // However, we handle this case in tenant loading code so the next time we attach, the issue is
+    // resolved.
```

Comment:
Ack, I don't think this can be avoided given that S3 isn't transactional. At least it'll get corrected on the next manifest upload (the next pageserver restart at the latest). Thanks for pointing it out.

Reply:
we could in theory add a `deleted` flag also to the tenant manifest, and then once we encounter it, we resume a deletion as well. The `deleted` flag would be set before the `index-part.json`.

But IIRC John wants to transition away from reliable deletions in the pageserver towards wanting control plane to retry deletions. So that's why I wouldn't spend too much time on it.